### ISSUE TITLE
regression: ast gen fails to catch incorrect by ref error captures

### ISSUE THREAD

**Issue by travisstaloch** [ziglang/zig#18583]
### Zig Version

0.12.0-dev.2158+4f2009de1

### Steps to Reproduce and Observed Behavior

this is a follow up to #18579 and addresses some changes made in #18173.  

these code snippets below were produced by @Techatrix and i think he may address them soon.  we noticed this while working on #18579 but decided that it deserves its own issue.

with 0.11.0 the following produced a compile error `error set cannot be captured by reference`.  but now the compiler allows it.  i think this is a regression.  
```zig
comptime {
    const e: error{A}!u32 = error.A;

    e catch |err| switch (err) {
        error.A => |*foo| {
            @compileLog(foo);
        },
    };
}
// 
```

The compiler also doesn't seem to catch the following syntax error:
```zig
comptime {
    const e: error{A}!u32 = error.A;

    if (e) |_| {} else |err| switch (err + "I can put anything I want here") {
        error.A => {},
    }
}
```

This is a little beyond me so I'm hoping that @Techatrix or another contributor knows what needs to be done here.

### Expected Behavior

should be a compile error like it was in 0.11

**Comment by travisstaloch** [comment]
i forgot to demonstrate that this used to be an error in 0.11.0 but now triggers an assertion w/ a debug build of the compiler.  

```zig
// /tmp/tmp.zig
test {
    const e: error{A}!u32 = error.A;
    e catch |err| switch (err) {
        error.A => |*foo| {
            _ = foo;
        },
    };
}
```

```console
/tmp $ zig version
0.11.0
/tmp $ zig test /tmp/tmp.zig
/tmp/tmp.zig:38:21: error: error set cannot be captured by reference
        error.A => |*foo| {
                    ^~~~
/tmp $ zigup master-build # switch to local current debug compiler build
...
/tmp $ zig version
0.12.0-dev.2237+e09450214
/tmp $ zig test /tmp/tmp.zig
thread 498506 panic: reached unreachable code
~/dev/zig/zig/lib/std/debug.zig:403:14: 0x172eaac in assert (zig)
    if (!ok) unreachable; // assertion failure
             ^
~/dev/zig/zig/src/AstGen.zig:7221:19: 0x253aa16 in switchExprErrUnion (zig)
            const capture_token = case.payload_token orelse break :blk &err_scope.base;
                  ^
~/dev/zig/zig/src/AstGen.zig:1034:50: 0x1fc22d4 in expr (zig)
                        return switchExprErrUnion(gz, scope, ri.br(), node, .@"catch");
                                                 ^
~/dev/zig/zig/src/AstGen.zig:2518:41: 0x2a205fc in unusedResultExpr (zig)
    const maybe_unused_result = try expr(gz, scope, .{ .rl = .none }, statement);
                                        ^
~/dev/zig/zig/src/AstGen.zig:2499:65: 0x2a23197 in blockExprStmts (zig)
                else => noreturn_src_node = try unusedResultExpr(gz, scope, inner_node),
                                                                ^
~/dev/zig/zig/src/AstGen.zig:2310:27: 0x255f73a in blockExpr (zig)
        try blockExprStmts(&block_scope, &block_scope.base, statements);
                          ^
~/dev/zig/zig/src/AstGen.zig:1007:33: 0x1fc4a8a in expr (zig)
                return blockExpr(gz, scope, ri, node, statements[0..2]);
                                ^
~/dev/zig/zig/src/AstGen.zig:4762:34: 0x2523dc6 in testDecl (zig)
    const block_result = try expr(&fn_block, &fn_block.base, .{ .rl = .none }, body_node);
                                 ^
~/dev/zig/zig/src/AstGen.zig:5669:28: 0x1fb8787 in containerMember (zig)
            astgen.testDecl(gz, scope, wip_members, member_node) catch |err| switch (err) {
                           ^
~/dev/zig/zig/src/AstGen.zig:4981:49: 0x1c75e32 in structDeclInner (zig)
        var member = switch (try containerMember(&block_scope, &namespace.base, &wip_members, member_node)) {
                                                ^
~/dev/zig/zig/src/AstGen.zig:165:35: 0x19c5944 in generate (zig)
        if (AstGen.structDeclInner(
                                  ^
~/dev/zig/zig/src/Module.zig:2839:35: 0x20d395b in astGenFile (zig)
    file.zir = try AstGen.generate(gpa, file.tree);
                                  ^
~/dev/zig/zig/src/Compilation.zig:3824:19: 0x1dbeb6e in workerAstGenFile (zig)
    mod.astGenFile(file) catch |err| switch (err) {
                  ^
~/dev/zig/zig/lib/std/Thread/Pool.zig:94:39: 0x1af0b75 in runFn (zig)
            @call(.auto, func, closure.arguments);
                                      ^
~/dev/zig/zig/lib/std/Thread/Pool.zig:133:18: 0x202f527 in worker (zig)
            runFn(&run_node.data);
                 ^
~/dev/zig/zig/lib/std/Thread.zig:411:13: 0x1d2656a in callFn__anon_81706 (zig)
            @call(.auto, f, args);
            ^
~/dev/zig/zig/lib/std/Thread.zig:684:30: 0x1a55822 in entryFn (zig)
                return callFn(f, args_ptr.*);
                             ^
./nptl/pthread_create.c:442:8: 0x745f11e94ac2 in start_thread (pthread_create.c)
../sysdeps/unix/sysv/linux/x86_64/clone3.S:81:0: 0x745f11f2684f in ??? (../sysdeps/unix/sysv/linux/x86_64/clone3.S)
???:?:?: 0x0 in ??? (???)
Aborted (core dumped)
```

**Comment by andrewrk** [comment]
@dweiller if you can fix this as a follow-up to #18173 then it can stay, otherwise I'm simply going to revert those changes and move forward with the higher priority 0.12.0 issues.

**Comment by dweiller** [comment]
> @dweiller if you can fix this as a follow-up to #18173 then it can stay, otherwise I'm simply going to revert those changes and move forward with the higher priority 0.12.0 issues.

Doesn't https://github.com/ziglang/zig/pull/18584 fix this?

**Comment by andrewrk** [comment]
Yes it does, great!
